Pull request 2

vitsalis opened the pull request.

vitsalis committed:
Refactor redirect exceptions and user::create, introduce go function

vitsalis committed:
Does some corrections

vitsalis committed:
Add whitespace

vitsalis committed:
Adds a link which redirects to index.php at profile

vitsalis committed:
See warnings, public function -> public static function

vitsalis committed:
Test commit

vitsalis committed:
Deleted some statics by mistake

vitsalis committed:
Adds active record on models

vitsalis committed:
Deletes .swn extention

vitsalis committed:
Does some corrections in the active record

vitsalis committed:
Did some corrections in the active record model at images

vitsalis committed:
Add config-local to allow for local secrets

vitsalis committed:
Adds active records on models/user.php

vitsalis committed:
Removes a wrong include

pkakelas committed:
replaced mail.php with php default email-validation function

pkakelas committed:
Added database schema

vitsalis committed:
Added ActiveRecordBase for users

pkakelas committed:
Pulled from VitSalis

pkakelas committed:
replaced mail validator with the default php function.

pkakelas committed:
adder a ) that I had forgotten, and updated db_schema.sql file

vitsalis committed:
Adds buggy version of active record on images

pkakelas committed:
Added population & migration script for countries

pkakelas committed:
Corrected migration script

pkakelas committed:
migration script

pkakelas committed:
perfected migration script

pkakelas committed:
perfected migration script

pkakelas committed:
Finished population.php

pkakelas committed:
new db_schema, changes to migration.php

vitsalis committed:
Not full changes

vitsalis committed:
Forgot to add a folder

vitsalis committed:
Merge branch 'master' of https://github.com/pkakelas/endofcodes into pull

vitsalis committed:
Merged from pkakelas

vitsalis committed:
Adds html for the register page

vitsalis committed:
Adds the header of the page

vitsalis committed:
Adds dropout box at header

vitsalis committed:
Adds html on login page and on home page

vitsalis committed:
Adds html for the home page

vitsalis committed:
Adds some CSS and HTML to the header

vitsalis committed:
Modifies header's css

pkakelas committed:
Pulled from Vitsalis

pkakelas committed:
Added sample codes to config.php

pkakelas committed:
Created register.css and linked it to head in header.php

pkakelas committed:
Started css in register page by styling the 4 step divs

vitsalis committed:
Added function includeStyle

vitsalis committed:
Added and modified css files.
general.css contains some CSS rules about all the files.
home.css contains some CSS about the home file, and does some
organizing, although it is not the final stage of home.
In header.php I modified some id to add css
In home.php I added some ids to add css

vitsalis committed:
Merge branch 'master' of https://github.com/pkakelas/endofcodes

vitsalis committed:
Added class "error" which makes the color of errors red

vitsalis committed:
Adds CSS

vitsalis committed:
Fixed some errors

vitsalis committed:
Now when a user selects a country it is inserted to the database

vitsalis committed:
Adds edit settings feature

vitsalis committed:
Changed the structure and the name of the db folder

vitsalis committed:
Changes db schema and adds an age migration script

vitsalis committed:
Deletes unnesecary information from db schema

pkakelas committed:
Moved countries_array.php to population folder and made Vitsalis corrections

pkakelas committed:
Moved countries_array.php to population folder and made Vitsalis corrections

vitsalis committed:
Merge branch 'master' into active_record

vitsalis committed:
Add db_select_one function for queries that return only one result

vitsalis committed:
Migrations are now numbered based on date

vitsalis committed:
Change images table field from imageid to id

vitsalis committed:
Fix image active record.

- Image has-a User object.
- Remove list of user-specific attributes from User
- getCurrentImage() converted into a finder factory
- Image constructor takes image id instaed of user id

vitsalis committed:
Replace all include with include_once

vitsalis committed:
Fix some buggs

vitsalis committed:
Remove .swn file

vitsalis committed:
Add .swn files to gitignore

vitsalis committed:
Removes unnesecary lines of code

vitsalis committed:
Removes interface with users table from image model

vitsalis committed:
Makes ActiveRecordBase an abstract class

dionyziz commented on database/population/countries_array.php:
```diff
@@ -0,0 +1,246 @@
+
+    function getCountries() {
+        return array( 
+            'AF'=>'Afghanistan',
```
Coding style.

vitsalis replied:
what's wrong?

dionyziz replied:
Spaces around fat arrow.

dionyziz commented on config/config.php:
```diff
@@ -1,4 +1,4 @@
-<?php
+ <?php
```
What?

vitsalis replied:
idk wtf, I will fix this

dionyziz commented on config/config.php:
```diff
@@ -16,4 +16,4 @@ function getConfig() {
         return $config;
     }
-?>
+?>                                                                                                          ?>
```
What?...

dionyziz commented on controllers/dashboard.php:
```diff
@@ -4,9 +4,9 @@ public static function view() {
-            include 'views/header.php';
-            include 'views/home.php';
-            include 'views/footer.php';
+            include_once 'views/header.php';
```
These should be includes, not include_once.

vitsalis committed:
Removed unnesecary spaces

vitsalis committed:
Renam some functions to include

dionyziz commented on controllers/image.php:
```diff
@@ -1,23 +1,27 @@
                 throw new HTTPUnauthorizedException();
             }
-            $username = $_SESSION[ 'user' ][ 'username' ];
+            $config = getConfig();
```
Consider making this a global variable.

dionyziz commented on controllers/image.php:
```diff
@@ -1,23 +1,27 @@
+            $image->tmp_name = $tmp_name;
+            $image->imagename = $imagename;
+            $image->user = $user;
+            $image->ext = Extention::get( $imagename );
```
Move this to within the model.

dionyziz commented on controllers/image.php:
```diff
@@ -1,23 +1,27 @@
             }
-            $username = $_SESSION[ 'user' ][ 'username' ];
+            $config = getConfig();
+            $imagename = basename( $image[ 'name' ] );
```
Move this to within the model.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,125 @@
-        public static function create( $username = '', $password = '', $email = '' ) {
-            include 'models/users.php';
-            include 'models/mail.php';
+        public static function create( $username = '', $password = '', $password_repeat = '', $email = '', $country, $accept = false ) {
```
Needs default value for $country.

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +85,34 @@
+            ?><p class="error">Please accept the terms of usage</p><?php
+        }
+    ?>
+    <p><input type="checkbox" name="accept" /> I agree on the <a href="">Terms of Usage</a></p>
```
I don't think we need terms right now. Let's get rid of this and make registration easier and simpler.

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +85,34 @@
         echo htmlspecialchars( $val );
     ?>"/></p>
+    <?php
+        if ( isset( $empty_country ) ) {
```
Let's not make this mandatory.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,125 @@
+            if ( empty( $password_repeat ) ) {
+                go( 'user', 'create', array( 'empty_pass_repeat' => true ) );
+            }
+            if ( $country === 'Select Country' ) {
```
Validation should not depend on textual value. Check to see if country is really valid using a whitelist.

dionyziz commented on models/country.php:
```diff
@@ -0,0 +1,16 @@
+<?php
+    class Country {
```
Migrate to active record.

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        protected function validate() {
+            if ( strlen( $this->password ) <= 6 ) {
```
Min password length should be configurable, not hard-coded.

vitsalis committed:
Fixed coding style

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,125 @@
-            include 'views/user/view.php';
+            $country = Country::getCountryName( $user->countryid );
+            $config = getConfig();
+            $image = Image::find_by_user( $user );
```
Use a has-a relationship between User and Image:

$user->avatar

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,125 @@
+            $country = Country::getCountryName( $user->countryid );
+            $config = getConfig();
+            $image = Image::find_by_user( $user );
+            $target_path = $config[ 'paths' ][ 'avatar_path' ] . $image->id . $image->ext;
```
Move path construction to Image model.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,125 @@
-                if ( $password_new != $password_repeat ) {
-                    throw new RedirectException( 'index.php?resource=user&method=update&not_matched=yes' );
+            $user = new User( $_SESSION[ 'user' ][ 'id' ] );
+            if ( $user->authenticatesWithPassword( $password ) ) {
```
Don't require users to type their password to change any setting. Just require them to type their old password to set a new password.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,125 @@
+                if ( !empty( $email ) ) {
+                    $user->email = $email;
+                }
+                if ( $country !== 'Select Country' ) {
```
Whitelist validation for country.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,125 @@
+                    $user->email = $email;
+                }
+                if ( $country !== 'Select Country' ) {
+                    $user->countryid = Country::getCountryId( $country );
```
Prefer: User has-a Country and set $user->country to a Country object.

dionyziz commented on database/db_schema.sql:
```diff
@@ -0,0 +1,71 @@
+CREATE TABLE IF NOT EXISTS `images` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `userid` int(11) NOT NULL,
+  `imagename` varchar(200) COLLATE utf8_unicode_ci NOT NULL,
```
Use `name` for this?

dionyziz commented on database/db_schema.sql:
```diff
@@ -0,0 +1,71 @@
+  `salt` varchar(200) COLLATE utf8_unicode_ci NOT NULL,
+  `avatarid` int(11) NOT NULL,
+  `countryid` int(4) unsigned NOT NULL,
+  `age` int(3) unsigned NOT NULL,
```
Use date of birth.

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,29 @@
+    include_once '../../config/config-local.php';
+    include_once '../../models/database.php';
+
+    $res1 = mysql_query( "CREATE TABLE IF NOT EXISTS 
```
Coding style, leave new line after mysql_query(

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,29 @@
+            `shortname` text COLLATE utf8mb4_unicode_ci NOT NULL,
+            PRIMARY KEY (`id`)
+        ) 
+        ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci AUTO_INCREMENT=1"
```
remove AUTO_INCREMENT from migration script

vitsalis replied:
why?

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,29 @@
+        ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci AUTO_INCREMENT=1"
+    );
+
+    $res2 = mysql_query( "ALTER TABLE
```
coding style

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,29 @@
+<?php
+    include_once '../../config/config-local.php';
```
Use migrate.php; change migrate.php to accept multiple queries in an array.

dionyziz commented on database/migration/20131205173810_age.php:
```diff
@@ -0,0 +1,10 @@
+        'ALTER TABLE
+            users
+        ADD COLUMN
+            age int(3) unsigned NOT NULL;'
```
Use date of birth.

dionyziz commented on database/population/country.php:
```diff
@@ -0,0 +1,29 @@
+        $array[] = "INSERT INTO 
+                        `countries` (`id`, `country`, `shortname`) 
+                    VALUES 
+                        (NULL, '$value', '$key');";
```
Use prepared queries.

dionyziz commented on database/population/country.php:
```diff
@@ -0,0 +1,29 @@
+    $keys = array_keys( $countries );
+  
+    foreach ( $keys as $key ) {
+        $value = $countries[$key];
```
style

dionyziz commented on models/base.php:
```diff
@@ -0,0 +1,35 @@
+
+    class ModelNotFoundException extends Exception {
+        public function __construct() {
+            parent::__construct( 'NotFound' );
```
Use a more descriptive error message: Maybe "Model not found"? Recall that this is seen by someone catching a generic exception.

dionyziz commented on models/base.php:
```diff
@@ -0,0 +1,35 @@
+
+    class ModelValidationException extends Exception {
+        public $error;
+        public function __construct( $error = "" ) {
```
Style: empty line between attributes and methods.

dionyziz commented on models/db.php:
```diff
@@ -70,7 +70,7 @@ function db_delete( $table, $where ) {
         }
         db(
-            'INSERT FROM '
+            'DELETE FROM '
```
Good catch!

dionyziz commented on models/dependencies.php:
```diff
@@ -1,6 +1,14 @@
+    include_once 'models/database.php';
+    include_once 'models/base.php';
     session_start();
+    error_reporting( E_ALL );
```
Move to header.php? Try to keep dependencies.php clear and have only file includes in it. All initialization should be in header.

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,66 @@
-        public static function create( $username, $tmp_name, $imagename, $userid ) {
-            global $config;
-            $ext = Extention::get( $imagename ); 
+    include_once 'models/user.php';
```
Style: Blank line between dependencies and code.

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,66 @@
+    class Image extends ActiveRecordBase {
+        public $user;
+        public $tmp_name;
+        public $imagename;
```
Consider renaming attribute to $name so that we can do $image->name instead of $image->imagename. I know that we use $user->username, but "username" is a word on its own so it makes sense in that context.

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,66 @@
+        public $user;
+        public $tmp_name;
+        public $imagename;
+        public $avatarid;
```
Why do we need this?

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,66 @@
+            global $config;
+
+            $tmp_name = $this->tmp_name;
+            $imagename = $this->imagename;
```
basename here

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,66 @@
+        public function upload() {
+            $tmp_name = $this->tmp_name;
+            $target_path = $this->target_path;
             return move_uploaded_file( $tmp_name, $target_path );
```
This might overwrite an existing image. Let's just use the database id for filenames instead.

vitsalis replied:
$target_path is 'avatars/' . $id . '.' . $ext
Check it here https://github.com/VitSalis/endofcodes/blob/master/models/image.php#L49

dionyziz replied:
OK, makes sense. Looks good, thank you for clarifying this.

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,109 @@
+<?php
+    include_once 'encrypt.php';
```
Style: Blank line after includes and before class definition/code.

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,109 @@
+        protected $tableName = 'users';
+
+        public static function find_by_username( $username ) {
+            $users = db_select( 'users', array( 'id' ), compact( "username" ) );
```
db_select_one

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,109 @@
+                $user_info = db_select_one( 'users', array( 'username', 'email', 'password', 'countryid', 'avatarid' ), compact( "id" ) );
+                $this->username = $user_info[ 'username' ];
+                $this->email = $user_info[ 'email' ];
+                $this->password = $user_info[ 'password' ];
```
This does not need to be selected from the database nor retrieved into an attribute at this point, as it cannot be read (it's encrypted).

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,109 @@
+                'users', 
+                compact( "username", "password", "email", "salt", "countryid" )
+            );
+            if ( $res === false ) { 
```
Use exceptions for failed queries.

vitsalis replied:
Can you help me do that?

dionyziz commented on views/header.php:
```diff
@@ -3,5 +3,50 @@
     <head>
         <title>EndofCodes Demo</title>
+        <?php
+            function includeStyle( $path ) {
```
Move to helpers/html.php.

dionyziz commented on views/header.php:
```diff
@@ -3,5 +3,50 @@
+            includeStyle( "register" );
+            includeStyle( "login" );
+        ?>
+        <meta 
```
Use HTML5 UTF8 meta tag.

dionyziz commented on views/header.php:
```diff
@@ -3,5 +3,50 @@
 <html>
```
Define language to be English.

dionyziz commented on views/header.php:
```diff
@@ -3,5 +3,50 @@
+                            echo htmlspecialchars( $_SESSION[ 'user' ][ 'username' ] );
+                        ?><select>
+                            <option> </option>
+                            <option value="logout">Logout</option>
```
This should be a form with a button or a link, not a select/option. This is semantically incorrect. If you want it to look like a drop-down, use CSS + JS.

vitsalis replied:
Yes you are right, it was stupid to mix presentation with content.

dionyziz commented on views/session/create.php:
```diff
@@ -1,23 +1,35 @@
 ?>
+<div id="login"><?php
+        if ( isset( $empty_user ) ) {
+            ?><p class="error">Please fill the username form.</p><?php
```
"Form" is textually incorrect; the form is the whole set of inputs. Use something clearer, such as "Please type a username". Similarly for below.

dionyziz commented on views/session/create.php:
```diff
@@ -1,23 +1,35 @@
+        <p><a href="">Forgot password?</a></p>
+        <p><input type="submit" value="Login" /></p>
+    </form>
+    <p><a href="index.php?resource=user&method=create">Don't have an account?</a></p>
```
HTML entity

dionyziz commented on views/user/create.php:
```diff
@@ -1,21 +1,44 @@
+        <p>Step 1</p>
+        <p><a href="">Download the libraries</a></p>
+    </div>
+     <div id="step2" class="steps">
```
style

dionyziz commented on views/user/create.php:
```diff
@@ -1,21 +1,44 @@
+    </div>
+     <div id="step2" class="steps">
+        <p>Step 2</p>
+        <p>Read the basic <a href="">rules</a></p>
```
The whole step has the rules, why make this word specifically a link?

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +85,34 @@
+    <select name="country">
+        <option>Select Country</option>
+        <?php
+            include_once 'db/country/countries_array.php';
```
Database scripts must not be used in production code!

dionyziz commented on views/user/update.php:
```diff
@@ -1,30 +1,54 @@
+    <label for="email">Email</label>
+    <?php
+        if ( isset( $mail_notvalid ) ) {
+            ?><p class="error">Mail is not valid</p><?php
```
Be consistent about the spelling of "Email". Use "Email" everywhere.

dionyziz commented on views/user/update.php:
```diff
@@ -1,30 +1,54 @@
+    <p><select name="country">
+        <option>Select Country</option>
+        <?php
+            include_once 'db/country/countries_array.php';
```
Same as above.

dionyziz commented on views/user/view.php:
```diff
@@ -25,14 +15,37 @@
             ?>
             <p><input type="file" name="image" id="image" /></p>
             <input type="submit" value="Upload" />
```
Move this to separate "edit profile" or "settings" page.

dionyziz commented on views/user/view.php:
```diff
@@ -25,14 +15,37 @@
+<p><img src="static/images/twitter-logo.png" alt="twitter link" width="40" height="40" /></p>
+<p><img src="static/images/github-logo.png" alt="github link" width="40" height="40" /></p>
+<p><img src="static/images/google+-logo.jpeg" alt="google+ link" width="40" height="40" /></p>
+<p><a href="">Add friend</a></p>
```
Create trello TODOs for all these. Don't check in code that doesn't work unless you create a relevant TODO.

dionyziz commented on views/user/view.php:
```diff
@@ -25,14 +15,37 @@
+<?php 
+    if ( $_SESSION[ 'user' ][ 'id' ] == $user->id ) {
+        ?><form action="index.php?resource=user&amp;method=delete" method="post">
+            <input type="submit" value="Delete your account" />
```
Move to settings page.

dionyziz commented on views/user/view.php:
```diff
@@ -25,14 +15,37 @@
+        ?><form action="index.php?resource=user&amp;method=delete" method="post">
+            <input type="submit" value="Delete your account" />
+        </form>
+        <p><a href="index.php?resource=user&method=update">Edit Settings</a></p><?php
```
entity

pkakelas committed:
pulled from vit

vitsalis committed:
Fix path to countries array

vitsalis committed:
Debugs images controller and model

- Change every avatarid name to id
- Do the nessecary changes for it to work

vitsalis committed:
Fixes a password bugg

pkakelas committed:
Merge branch 'master' of https://github.com/VitSalis/endofcodes

pkakelas committed:
Added dob selects in user create view and created month_array.php in population folder

pkakelas committed:
coding style

pkakelas committed:
coding style

pkakelas committed:
added dob restrictions in user controller

pkakelas committed:
Made dob restrictions workable

pkakelas committed:
Added dob in the database by configuring user model and controller

pkakelas committed:
Changed dob in database to have date format

pkakelas committed:
added validation to username input

vitsalis committed:
Give changedPass a default value

vitsalis committed:
Remove database connection with views

pkakelas committed:
coding style

pkakelas committed:
Changed country migration to work with migrate function

vitsalis committed:
Move functions from dependencies to header for architecture reasons

vitsalis committed:
Fix style

vitsalis committed:
Rename column imagename to name

vitsalis committed:
Rename all variables imagename to name

vitsalis committed:
Forgot to add basename at image->tmp_name

vitsalis committed:
Fix style

vitsalis committed:
db_select_one and $users becomes $user

vitsalis committed:
Remove password from construct

vitsalis committed:
Move function includeStyle to helpers

vitsalis committed:
Define language to be english

vitsalis committed:
Use HTML5 utf8 meta tag

vitsalis committed:
removed for now dropdown box near username

vitsalis committed:
Fix lectical error

vitsalis committed:
Add html entities

vitsalis committed:
Move delete account form to settings page

vitsalis committed:
Move upload picture to edit settings page

vitsalis committed:
Views doesn't interface with database anymore

vitsalis committed:
Use the name email instead of mail everywhere

vitsalis committed:
Make the whole sentence a link

vitsalis committed:
Fix style

vitsalis committed:
Fix style

vitsalis committed:
Add a more descriptive message at ModelNotFoundException

vitsalis committed:
Fix style

vitsalis committed:
Use fixed queries for population script

vitsalis committed:
Pulled from pkakelas

vitsalis committed:
Add migration script to delete age and add dob

vitsalis committed:
Debuged code

vitsalis committed:
Changed migrate to accept multiple sql statements

vitsalis committed:
Adds country whitelist and doesn't check values by string

vitsalis committed:
Now you need to know your password only to change it

vitsalis committed:
Target path is now made at image model

vitsalis committed:
Now min password len isn't hard coded

vitsalis committed:
Did some architecture changes.

-Now it isn't mandatory to select a country or dob.
-Deleted the accept the terms checkbox.
-Instead of deleting the checkers I commented them, in case of need.

vitsalis committed:
Move basename to model

vitsalis committed:
Model finds extention now

vitsalis committed:
Make config a global variable

vitsalis committed:
Deletes useless variable

pkakelas committed:
Changed database schema

pkakelas committed:
Migrates country.php to active_Record

pkakelas committed:
Added all country flags

vitsalis committed:
fuck you

vitsalis committed:
Revert "Deletes useless variable"

This reverts commit 4fa68251fbfd7ac0fd20b54f76f1ffa494cab9d2.

vitsalis committed:
Reverting a stupid mistake

vitsalis committed:
Merge branch 'master' of https://github.com/pkakelas/endofcodes into debug

vitsalis committed:
Fix active_Record in image

vitsalis committed:
Merge branch 'master' of https://github.com/pkakelas/endofcodes

vitsalis committed:
Merged from pkakelas

vitsalis committed:
Merge branch 'master' into debug

vitsalis committed:
Now user has-a country and has-a image

dionyziz commented on controllers/image.php:
```diff
@@ -1,23 +1,26 @@
                 throw new HTTPUnauthorizedException();
             }
-            $username = $_SESSION[ 'user' ][ 'username' ];
+            $name = $image[ 'name' ];
```
Why do we need extra variables here? Seems like direct assignment may be clearer and shorter in this case. Same for line below.

dionyziz commented on controllers/image.php:
```diff
@@ -1,23 +1,26 @@
             try {
-                Image::create( $username, $tmp_name, $avatarname, $id );
+                $user->image->save();
+                $user->save();
```
Why are you saving the user? $user->image->save() should do everything needed to update the avatar.

dionyziz commented on controllers/session.php:
```diff
@@ -1,31 +1,37 @@
+                go( 'session', 'create', array( 'wrong_pass' => true ) );
+            }
+            $id = $user->id;
             $_SESSION[ 'user' ] = array(
```
compact?

dionyziz commented on controllers/session.php:
```diff
@@ -1,31 +1,37 @@
-        public static function createView( $error, $empty_user, $empty_pass ) {
-            include 'views/session/create.php';
+        public static function createView( $wrong_pass, $empty_user, $empty_pass, $wrong_user ) {
+            include_once 'views/session/create.php';
```
Views should be used with include, not include_once

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
-        public static function create( $username = '', $password = '', $email = '' ) {
-            include 'models/users.php';
-            include 'models/mail.php';
+        public static function create( $username = '', $password = '', $password_repeat = '', $email = '', $countryname = '', /*$accept = false, */$day = '', $month = '', $year = '' ) {
```
line too long

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+        public static function create( $username = '', $password = '', $password_repeat = '', $email = '', $countryname = '', /*$accept = false, */$day = '', $month = '', $year = '' ) {
+            include_once 'models/user.php';
+            include_once 'models/country.php';
+            include_once 'database/population/months_array.php';
```
don't use database scripts in production code

vitsalis replied:
that's a controller..

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+            if ( empty( $username ) ) {
+                go( 'user', 'create', array( 'empty_user' => true ) );
+            }
+            if ( strpos( $username, ' ' ) || preg_match('#[^a-zA-Z0-9]#', $username)) {
```
style

dionyziz replied:
regexp doesn't work as expected, test it and fix it http://rubular.com/

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+            if ( empty( $password_repeat ) ) {
+                go( 'user', 'create', array( 'empty_pass_repeat' => true ) );
+            }
+            if ( !Country::onList( $countryname ) ) {
```
Use active record pattern:

```
try {
    Country::find_by_name( ... )
}
```

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+                $year = 0;
+                //go( 'user', 'create', array( 'empty_year' => true ) );
+            }
+            $month = array_search ($month, $months);
```
style

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+                $day = 0;
+                //go( 'user', 'create', array( 'empty_day' => true ) );
+            }
+            $months = getMonths();
```
Your date validation is laughable, use some real date validation. http://php.net/checkdate

dionyziz replied:
Move date validation to model.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+            $user->password = $password;
+            $user->email = $email;
+            $user->dob = $dob;
+            $user->countryid = $country->id;
```
Use a has-a relationship.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+                go( 'user', 'create', array( $e->error => true ) );
             }
-            throw new RedirectException( 'index.php?empty_mail=yes&resource=user&method=create' );
+            $_SESSION[ 'user' ] = array(
```
compact

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+            /*if ( $accept === false ) {
+                go( 'user', 'create', array( 'not_accepted' => true ) );
+            }*/
+            if ( empty( $username ) ) {
```
move username/password/email emptiness validation to model

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+                    go( 'user', 'update', array( 'wrong_pass' => true ) );
+                }
+            }
+            else if ( !empty( $password_new ) || !empty( $password_repeat ) ) {
```
If these aren't empty, only then check if user authenticates with password $password. Merge with above if case. No need to check if $password is empty.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+            if ( !empty( $email ) ) {
+                $user->email = $email;
+            }
+            if ( Country::onList( $countryname ) ) {
```
Use active record pattern as above.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+                $user_used, $small_pass, $mail_used, $mail_notvalid/*, $empty_country, $not_accepted, $empty_day, $empty_month, $empty_year*/ ) {
+            include_once 'models/country.php'; 
+            $countries = Country::getAll();
+            include_once 'views/user/create.php';
```
include

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+        public static function createView( $empty_user, $invalid_username, $empty_mail, $empty_pass, $empty_pass_repeat, $not_matched,
+                $user_used, $small_pass, $mail_used, $mail_notvalid/*, $empty_country, $not_accepted, $empty_day, $empty_month, $empty_year*/ ) {
+            include_once 'models/country.php'; 
+            $countries = Country::getAll();
```
find_all

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
-            include 'views/user/update.php';
+        public static function updateView( $notvalid, $small_pass, $not_matched, $wrong_pass, $mail_notvalid, $mail_used, $empty_country  ) {
+            include_once 'models/country.php';
+            $countries = Country::getAll();
```
find_all

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
 
-        public static function updateView( $small_pass, $not_matched, $old_pass ) {
-            include 'views/user/update.php';
+        public static function updateView( $notvalid, $small_pass, $not_matched, $wrong_pass, $mail_notvalid, $mail_used, $empty_country  ) {
```
Use consistent names for errors: Maybe fieldname_problem? "notvalid" is a bad name. "not_matched" is a bad name. Better names:

username_invalid
password_small
password_mismatch
password_wrong
email_invalid
email_used

etc.

dionyziz commented on controllers/user.php:
```diff
@@ -1,84 +1,154 @@
+        public static function updateView( $notvalid, $small_pass, $not_matched, $wrong_pass, $mail_notvalid, $mail_used, $empty_country  ) {
+            include_once 'models/country.php';
+            $countries = Country::getAll();
+            include_once 'views/user/update.php';
```
include

dionyziz commented on database/db_schema.sql:
```diff
@@ -0,0 +1,71 @@
+--
+
+CREATE TABLE IF NOT EXISTS `images` (
+  `imageid` int(11) NOT NULL AUTO_INCREMENT,
```
id

dionyziz commented on database/db_schema.sql:
```diff
@@ -0,0 +1,71 @@
+CREATE TABLE IF NOT EXISTS `images` (
+  `imageid` int(11) NOT NULL AUTO_INCREMENT,
+  `userid` int(11) NOT NULL,
+  `imagename` varchar(200) COLLATE utf8_unicode_ci NOT NULL,
```
name

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,18 @@
+    include_once 'migrate.php';
+
+    $sql1 = "CREATE TABLE IF NOT EXISTS 
+        `countries` (
```
style: one level of indentation deeper than above line within quotes

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,18 @@
+        ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci AUTO_INCREMENT=1";
+
+    $sql2 = "ALTER TABLE
+            `users` 
```
same

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,18 @@
+    $sql2 = "ALTER TABLE
+            `users` 
+        ADD COLUMN
+            `countryid` int(4) unsigned NOT NULL;";
```
Type must match type of countries.id

dionyziz commented on database/migration/20131126210913_country.php:
```diff
@@ -0,0 +1,18 @@
+            `users` 
+        ADD COLUMN
+            `countryid` int(4) unsigned NOT NULL;";
+    migrate( array( $sql1, $sql2 ) );
```
would put SQL directly here without the use of variables. Just use correct styling and plenty of new lines.

dionyziz commented on database/migration/20131205173810_age.php:
```diff
@@ -0,0 +1,9 @@
+            users
+        ADD COLUMN
+            age int(3) unsigned NOT NULL;';
+    migrate( array( $sql ) );
```
put SQL directly here

dionyziz commented on database/migration/20131207062100_change_imagesid_to_id.php:
```diff
@@ -0,0 +1,9 @@
+            images
+        CHANGE 
+            `imageid`  `id` INT( 11 ) NOT NULL AUTO_INCREMENT";
+    migrate( array( $sql ) );
```
direct SQL

dionyziz commented on database/migration/20131208144508_imagename_to_name.php:
```diff
@@ -0,0 +1,14 @@
+            images
+        ADD COLUMN
+            `name` text COLLATE utf8mb4_unicode_ci NOT NULL";
+    migrate( array( $sql1, $sql2 ) );
```
direct SQL

dionyziz commented on database/migration/20131208174017_age_delete.php:
```diff
@@ -0,0 +1,9 @@
+            users
+        DROP COLUMN
+            age';
+    migrate( array( $sql ) );
```
direct SQL

dionyziz commented on database/migration/20131208174309_add_dob.php:
```diff
@@ -0,0 +1,9 @@
+            users
+        ADD
+            dob date NOT NULL';
+    migrate( array( $sql ) );
```
same

dionyziz commented on database/population/countries_array.php:
```diff
@@ -0,0 +1,246 @@
+<?php
+
```
First line within <?php shouldn't be a blank line. We don't use that anywhere else. Be consistent.

dionyziz commented on database/migration/migrate.php:
```diff
@@ -0,0 +1,14 @@
+        include_once '../../models/database.php';
+
+        foreach ( $sql_array as $sql ) {
+            $res = mysql_query( $sql );
```
Let's use db() here?

dionyziz commented on database/population/country.php:
```diff
@@ -0,0 +1,23 @@
+    include_once '../../models/db.php';
+    include_once 'countries_array.php';
+    
+    $countries = getCountries();
```
get_countries

vitsalis replied:
dude our styling is getCountries.. We have talked about this

dionyziz replied:
Then change everything to this style. You are not being consistent. Your User class has a "find_by_user", your db.php has a "db_select". Choose one style and stick with it, and change everything to use it.

dionyziz commented on database/population/country.php:
```diff
@@ -0,0 +1,23 @@
+    $count = 0;
+    $keys = array_keys( $countries );
+  
+    foreach ( $keys as $key ) {
```
foreach ( $countries as $key => $value )

dionyziz commented on database/population/country.php:
```diff
@@ -0,0 +1,23 @@
+  
+    foreach ( $keys as $key ) {
+        $value = $countries[ $key ];
+        $res = db_insert( 'countries', array( 'id' => NULL, 'country' => $value, 'shortname' => $key ) );
```
id column is not needed

dionyziz commented on database/population/country.php:
```diff
@@ -0,0 +1,23 @@
+    foreach ( $keys as $key ) {
+        $value = $countries[ $key ];
+        $res = db_insert( 'countries', array( 'id' => NULL, 'country' => $value, 'shortname' => $key ) );
+        if ( $res === false ) { 
```
should throw exception on error

dionyziz commented on database/population/country.php:
```diff
@@ -0,0 +1,23 @@
+        ++$count;
+    }
+    
+    echo "You imported $count raws out of the 239 countries in the table 'countries'.";
```
rows

dionyziz commented on database/population/months_array.php:
```diff
@@ -0,0 +1,20 @@
+<?php
```
http://php.net/jdmonthname

vitsalis replied:
I don't know how to use that..

dionyziz replied:
Remove this file completely, loop from 1 to 12 where used, just call jdmonthname($i, 0);

dionyziz commented on models/country.php:
```diff
@@ -0,0 +1,35 @@
+        public $id;
+        public $shortname;
+
+        public static function getByName( $country ) {
```
find_by_name

dionyziz commented on models/country.php:
```diff
@@ -0,0 +1,35 @@
+            }
+        }
+
+        public static function getAll() {
```
find_all

dionyziz replied:
move static functions to top of class

dionyziz commented on models/country.php:
```diff
@@ -0,0 +1,35 @@
+        public function __construct( $id = false ) {
+            if ( $id ) {
+                $row = db_select_one( "countries", array( 'country', 'shortname' ), compact( "id" ) );
+                $this->name = $row[ 'country' ];
```
model attribute names must match table attribute names

dionyziz commented on models/country.php:
```diff
@@ -0,0 +1,35 @@
+            return db_select( 'countries' );
+        }
+
+        public static function onList( $country ) { 
```
remove and use exceptions

dionyziz commented on models/db.php:
```diff
@@ -79,22 +79,25 @@ function db_delete( $table, $where ) {
     }
 
+    function db_select_one( $table, $select = array( "*" ), $where = array() ) {
+        return array_pop( db_select( $table, $select, $where ) );
```
Consider throwing exception if no rows are returned (db() should also throw exception on error)

vitsalis replied:
what do you mean?

dionyziz replied:
If db_select returns 1 row, return it. If it returns multiple rows or 0 rows, throw a DBException here. Don't return false/null/empty array.

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,64 @@
+                $this->exists = false;
+            }
+            else {
+                $image_info = db_select_one( 'images', array( '*' ), compact( "id" ) );
```
must set exists to true

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,64 @@
+
+        public function __construct( $id = false ) {
+            if ( !$id ) {
+                $this->exists = false;
```
false is default value, no need to set

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,64 @@
+        public $ext;
+        public $userid;
+        protected $tableName = 'images';
+        protected $exists;
```
already specified in parent class

dionyziz commented on models/base.php:
```diff
@@ -0,0 +1,36 @@
+<?php
+    abstract class ActiveRecordBase {
```
add exists attribute common to all active record classes

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,64 @@
+                $this->id = $id;
+                $this->name = $image_info[ 'name' ];
+                $this->ext = Extention::get( $this->name );
+                $config = getConfig();
```
Make config a global variable.

vitsalis replied:
I tried adding a global $config, but still I need to define it in other classes..

dionyziz replied:
Not sure what the issue is, catch me on FB to discuss this.

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,64 @@
-                return "$id" . "." . $ext;
-            }
-            return false;
+            $name = "$id" . "." . $ext;
```
No need for quotes around $id.

dionyziz commented on models/image.php:
```diff
@@ -1,54 +1,64 @@
+            $this->target_path = $target_path . $name;
+            $this->id = $id;
+            $this->upload();
         }
```
Must also save user.

dionyziz replied:
Must set exists to true at this point.

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+        public $country;
+        public $image;
+        public $salt;
+        protected $exists;
```
move to parent class

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+
+        public static function find_by_username( $username ) {
+            $user = db_select_one( 'users', array( 'id' ), compact( "username" ) );
+            if ( empty( $user ) ) {
```
use exception

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+        public function __construct( $id = false ) {
+            if ( $id === false ) {
+                // new active record object
+                $this->exists = false;
```
exists is false by default

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+                $this->email = $user_info[ 'email' ];
+                $this->country = new Country( $user_info[ 'countryid' ] );
+                $this->image = new Image( $user_info[ 'avatarid' ] );
+                $this->id = $this->image->id = $id;
```
what?

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        protected function validate() {
+            $password_length = 6;
```
move to config

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+
+        protected function validate() {
+            $password_length = 6;
+            if ( isset( $this->password ) && strlen( $this->password ) <= $password_length ) {
```
password must be set during creation, validate correctly

vitsalis replied:
yes but I have an update too..

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+                'users', 
+                compact( "username", "password", "email", "salt", "countryid", "dob" )
+            );
+            if ( $res === false ) { 
```
use a DBException

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+                compact( "email", "password", "salt", "countryid", "avatarid", "dob" ),
+                compact( "id" )
+            );
+            if ( $res === -1 ) {
```
use a DBException

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,117 @@
+                }
+            }
+            $this->exists = true;
+            $this->id = mysql_insert_id();
```
Use return value of db_insert

dionyziz commented on views/header.php:
```diff
@@ -1,7 +1,40 @@
+            includeStyle( "register" );
+            includeStyle( "login" );
+        ?>
+        <meta charset="utf-8"> 
```
close tag

dionyziz commented on views/user/create.php:
```diff
@@ -1,21 +1,47 @@
+            ?><p class="error">Please type a username.</p><?php
+        }
+        if ( isset( $invalid_username ) ) { 
+            ?><p class="error">Please insert a valid username</p><?php
```
Tell me what's wrong with the username.

vitsalis replied:
How?

dionyziz replied:
"Usernames must start with a zonar sign, have at least 30 billion letters, and must end with a space."

dionyziz commented on views/user/create.php:
```diff
@@ -32,18 +58,23 @@
     <?php
         if ( isset( $mail_used ) ) {
-            ?><p>Mail is already used</p><?php
+            ?><p class="error">Mail is already used</p><?php
```
"Email"

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +88,80 @@
+    <select name="month">
+        <option>Select Month</option>
+        <?php
+            include_once 'database/population/months_array.php';
```
Don't use database scripts in production.

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +88,80 @@
+    <select name="day">                                                                                      
+        <option>Select Day</option>
+        <?php
+            for ( $i = 1; $i <= 31; $i++ ) {
```
Style ++$i

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +88,80 @@
+    <select name="year">
+        <option>Select Year</option>
+        <?php
+            for ( $i = 2007; $i >= 1910; $i-- ) {
```
Style --$i

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +88,80 @@
     <p><input type="text" id="email" name="email" value="<?php
         echo htmlspecialchars( $val );
     ?>"/></p>
+    <select name="day">                                                                                      
```
Add "Date of birth" label.

dionyziz commented on views/user/create.php:
```diff
@@ -57,9 +88,80 @@
+        ?>
+    </select> 
+    <?php
+        /*if ( isset( $empty_day ) ) {
```
Why are these commented out?

vitsalis replied:
because you said, typing a country or an age shouldn't be mandatory, but I am saving them for future use.

dionyziz commented on views/user/update.php:
```diff
@@ -1,30 +1,70 @@
 
 <?php
-    include 'views/footer.php';
+    include_once 'views/footer.php';
```
include

dionyziz commented on views/user/view.php:
```diff
@@ -1,38 +1,36 @@
 <?php
-    include 'views/header.php';
+    include_once 'views/header.php';
```
include, everywhere in views

dionyziz commented on views/user/view.php:
```diff
@@ -1,38 +1,36 @@
-    }
-?>
+<p>Region, Country: <?php
+    echo $user->country->name;
```
htmlspecialchars

dionyziz commented on views/header.php:
```diff
@@ -1,7 +1,40 @@
+        <div id="header">
+            <ul>
+                <li><a href="index.php">Endofcodes</a></li>
+                <li><a href="">Rules</a></li>
```
Needs trello TODO

vitsalis committed:
Delete extra variables

vitsalis committed:
Add compact

vitsalis committed:
include not include_once on views

vitsalis committed:
Make code more beautiful

vitsalis committed:
Move password, mail, username emptiness validation at models

vitsalis committed:
Removed emptiness of password repeat

vitsalis committed:
Fixes some buggs

vitsalis committed:
Refactor user/create and find country more like the active record way

vitsalis committed:
I now use checkdate

vitsalis committed:
Compact instead of array

vitsalis committed:
Fix code architecture

vitsalis committed:
Makes controller more active record style and fixes bug

vitsalis committed:
Rename function getAll to findAll

vitsalis committed:
Changed some names to better ones

vitsalis committed:
Renamed function getAll to findAll

vitsalis committed:
Updated db_schema

vitsalis committed:
Insert directly sql in migrate function

vitsalis committed:
Made country.id the same type as user.countryid

vitsalis committed:
I use db() instead of mysql_array

vitsalis committed:
Fix style

vitsalis committed:
Fix foreach to be architectical correct

vitsalis committed:
Removed collumn id because it wasn't needed

vitsalis committed:
Fix grammatical error

vitsalis committed:
Removed extra spaces

vitsalis committed:
Variable exists is now at ActiveRecordBase class

vitsalis committed:
Renamed getByName to findByName

vitsalis committed:
Renamed country.country to country.name at database

vitsalis committed:
Remove useless function and move static methods on top

vitsalis committed:
Fix architecture

vitsalis committed:
Make global the config variable

vitsalis committed:
Removed quotes around variable

vitsalis committed:
Remove useless if

vitsalis committed:
Add password length to config

vitsalis committed:
Do some debugging

vitsalis committed:
Use $res as a user id

vitsalis committed:
Add a forgoten close tag

vitsalis committed:
mail->email

vitsalis committed:
added a label and paragraphs

vitsalis committed:
fix style

vitsalis committed:
removed database from views

vitsalis committed:
htmlspecialchars added

vitsalis committed:
Add DBException

vitsalis committed:
Add DBException in models instead of ifs

dionyziz commented on config/config.php:
```diff
@@ -12,7 +12,8 @@ function getConfig() {
                 'avatar_path' => 'sample_path'
-            )
+            ),
+            'pass_len' => 'sample_pass_len'
```
Make this a sane default e.g. 6

dionyziz commented on database/population/countries_array.php:
```diff
@@ -0,0 +1,245 @@
+<?php
+    function getCountries() {
```
get_countries

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+            }*/
+            include_once 'models/user.php';
+            include_once 'models/country.php';
+            include_once 'database/population/months_array.php';
```
Do not ever use anything within "database" within real code. These includes should not exist, ever anywhere. Code within the database folder should never run except for administration tasks. The project should function in production even if the database folder is deleted.

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+            if ( $password !== $password_repeat ) {
+                go( 'user', 'create', array( 'not_matched' => true ) );
+            }
+            $months = getMonths();
```
Remove, use for loop.

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+                $country->name = '';
+                $country->id = 0;
+            }
             $_SESSION[ 'create_post' ] = array(
```
compact

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,128 @@
+            }
+        }
+
+        protected function validate() {
```
must validate dob

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        protected function validate() {
+            $config = getConfig();
```
use global

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+            }
+            $months = getMonths();
+            $month = array_search( $month, $months );
+            if ( is_string( $day ) || is_string( $month ) || is_string( $year ) ) {
```
Not sure what this is doing.

vitsalis replied:
I don't know how else to check if user has put a date..

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
 
-        public static function update( $password_old, $password_new, $password_repeat ) {
-            include 'models/users.php';
+        public static function update( $password = '', $password_new = '', $password_repeat = '', $countryname = '', $email = '' ) {
```
Countries should be identified by primary key (id) at the code level, not by name.

vitsalis replied:
But the form is giving the countryname not the id...

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+            $user = new User( $_SESSION[ 'user' ][ 'id' ] );
+            if ( !empty( $password_new ) || !empty( $password_repeat ) ) {
+                if ( $user->authenticatesWithPassword( $password ) ) {
+                    if ( !empty( $password_new ) || !empty( $password_repeat ) ) {
```
You're already checking for this above.

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+                $user->email = $email;
+            }
+            try {
+                $user->country = Country::findByName( $countryname );
```
find_by_name

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
 
-        public static function createView( $empty_user, $empty_mail, $empty_pass, $user_used, $small_pass, $mail_used, $mail_notvalid ) {
+        public static function createView( $empty_user, $invalid_username, $empty_mail, $empty_pass, $empty_pass_repeat, $not_matched,
+                $user_used, $small_pass, $mail_used, $mail_notvalid/*, $empty_country, $not_accepted, $empty_day, $empty_month, $empty_year*/ ) {
```
Why are empty_day etc. commented out?

vitsalis replied:
because it isn't mandatory to put a country

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
         }
 
-        public static function createView( $empty_user, $empty_mail, $empty_pass, $user_used, $small_pass, $mail_used, $mail_notvalid ) {
+        public static function createView( $empty_user, $invalid_username, $empty_mail, $empty_pass, $empty_pass_repeat, $not_matched,
```
Use better variable names, see comment in previous diff.

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+        public static function createView( $empty_user, $invalid_username, $empty_mail, $empty_pass, $empty_pass_repeat, $not_matched,
+                $user_used, $small_pass, $mail_used, $mail_notvalid/*, $empty_country, $not_accepted, $empty_day, $empty_month, $empty_year*/ ) {
+            include_once 'models/country.php'; 
+            include_once 'database/population/months_array.php';
```
No includes from database, see previous comment.

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+                $user_used, $small_pass, $mail_used, $mail_notvalid/*, $empty_country, $not_accepted, $empty_day, $empty_month, $empty_year*/ ) {
+            include_once 'models/country.php'; 
+            include_once 'database/population/months_array.php';
+            $months = getMonths();
```
Use for loop, see previous comment.

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
+            include_once 'models/country.php'; 
+            include_once 'database/population/months_array.php';
+            $months = getMonths();
+            $countries = Country::findAll();
```
find_all

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
         }
 
-        public static function updateView( $small_pass, $not_matched, $old_pass ) {
+        public static function updateView( $image_notvalid, $small_pass, $pass_not_matched, $wrong_pass, $mail_notvalid, $mail_used, $empty_country  ) {
```
Use better variable names.

dionyziz commented on controllers/user.php:
```diff
@@ -1,83 +1,132 @@
-        public static function updateView( $small_pass, $not_matched, $old_pass ) {
+        public static function updateView( $image_notvalid, $small_pass, $pass_not_matched, $wrong_pass, $mail_notvalid, $mail_used, $empty_country  ) {
+            include_once 'models/country.php';
+            $countries = Country::findAll();
```
find_all

dionyziz commented on models/base.php:
```diff
@@ -0,0 +1,44 @@
+        }
+    }
+
+    class DBException extends Exception {
```
move to db.php

dionyziz commented on models/base.php:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    class DBException extends Exception {
+        public function __construct() {
```
include error description from mysql_error()

dionyziz commented on models/user.php:
```diff
@@ -0,0 +1,128 @@
+        protected function validate() {
+            $config = getConfig();
+            if ( empty( $this->username ) ) {
+                throw new ModelValidationException( 'empty_user' );
```
Use consistent error names, this is not good.

dionyziz commented on views/user/create.php:
```diff
@@ -57,6 +88,80 @@
+        </select> 
+    </p> 
+    <?php
+        /*if ( isset( $empty_day ) ) {
```
Why is this commented out?

vitsalis replied:
Because putting an age or a country isn't mandatory..

dionyziz commented on views/user/update.php:
```diff
@@ -3,26 +3,66 @@
+            ?><p class="error">Email is not valid</p><?php
+        }
+        if ( isset( $mail_used ) ) {
+            ?><p class="error">Email is used</p><?php
```
"Email is already used" may be clearer.

dionyziz commented on views/user/update.php:
```diff
@@ -3,26 +3,66 @@
+    <p><input type="submit" value="Save settings" /></p>
+</form>
+
+<form action="index.php?resource=image&amp;method=create" method="POST" enctype="multipart/form-data"> 
```
POST must be lower case here.

vitsalis committed:
Made passlen a constant

vitsalis committed:
Compact

vitsalis committed:
Remove useless if

vitsalis committed:
Change the name of some variables

vitsalis committed:
Fixed style of function names

vitsalis committed:
Removed any database interface in the controllers

vitsalis committed:
Moved DBexception to db.php

vitsalis committed:
Add an mysql_error message at the constructor

vitsalis committed:
If dbSelectOne returns more than one row throw DBException

vitsalis committed:
Fix a bugg and make config a global variable

vitsalis committed:
Fix a bugg

vitsalis committed:
Fix another bugg

vitsalis committed:
Better error message

vitsalis committed:
more clear erros message

vitsalis committed:
lower case post

vitsalis committed:
Fixed a months bug

vitsalis committed:
Fix a session bug

vitsalis committed:
Fix variable names

vitsalis committed:
Change <= to < and name

vitsalis committed:
I check now with checkdate

vitsalis committed:
Fixed a correct dob checker and made countries value their id

vitsalis committed:
Changed variable names to better ones

vitsalis committed:
Now mail is prefiled in update

vitsalis committed:
Remove a comment and make config global

vitsalis committed:
Remove useless function

vitsalis committed:
Add tableName variable

vitsalis committed:
DB exception constructs with mysql_error

vitsalis committed:
Fix style

vitsalis committed:
Remove basename to avoid possible erros

vitsalis committed:
Fix style

vitsalis committed:
Make tableName protected and correct a mistake

vitsalis committed:
Fix a bug and make finding months more beautiful
